### ISSUE TITLE
SMOKE Array Out of Bound

### ISSUE THREAD

**Issue by hnqtran** [CEMPD/SMOKE#85]
Several subroutines in SMOKE are found to have array out of bound issue when SMOKE was compiled with gfortran compiler. The issue was also detected if SMOKE was compiled with intel compiler and with flag "-check bounds" activate in Makeinclude file (which is not enabled in the standard distributed package)

List of subroutines that having the issue (will be updated as more of them are found):
+ src/emqa/bldrepidx.f
+ src/smkinven/wrpdemis.f
+ src/emqa/rdssup.f

Details on the issue in each of the above subroutines are their mitigations are discussed in the below sections.

**Comment by hnqtran** [comment]
**Details on issues with src/smkinven/wrpdemis.f**
(more details to be added)
```
forrtl: severe (408): fort: (3): Subscript #1 of the array UCASNKEP has value 0 which is less than the lower bound of 1

Image              PC                Routine            Line        Source
smkinven           00000000004DD153  wrpdemis_                 284  wrpdemis.f
smkinven           000000000041EB22  genpdout_                 302  genpdout.f
smkinven           00000000004AD513  MAIN__                    502  smkinven.f
smkinven           00000000004072DD  Unknown               Unknown  Unknown
libc-2.28.so       0000152D846CDD85  __libc_start_main     Unknown  Unknown
smkinven           00000000004071FE  Unknown               Unknown  Unknown
31.397u 0.353s 0:32.17 98.6%    0+0k 464+104328io 4pf+0w
```

**Comment by hnqtran** [comment]
**Details on issues with src/emqa/rdssup.f**

```
forrtl: severe (408): fort: (2): Subscript #1 of the array GSPROID has value 22 which is greater than the upper bound of 21

Image              PC                Routine            Line        Source
smkreport          0000000000490953  rdssup_                   219  rdssup.f
smkreport          00000000004555F9  rdrepin_                  708  rdrepin.f
smkreport          0000000000471F7E  MAIN__                    201  smkreport.f
smkreport          000000000040731D  Unknown               Unknown  Unknown
libc-2.28.so       000014D7F8E16D85  __libc_start_main     Unknown  Unknown
smkreport          000000000040723E  Unknown               Unknown  Unknown
```

[Code line where issue occurs](https://github.com/CEMPD/SMOKE/blob/97d0e2c759556b2a1da2489cee80814c6954921b/src/emqa/rdssup.f#L219C2-L219C56)

**Comment by cseppan** [comment]
Fixes in commit 5bb27fd67a3f464cd77b9cf9854d6496be801615

**Comment by hnqtran** [comment]
> **Details on issues with src/smkinven/wrpdemis.f** (more details to be added)
> 
> ```
> forrtl: severe (408): fort: (3): Subscript #1 of the array UCASNKEP has value 0 which is less than the lower bound of 1
> 
> Image              PC                Routine            Line        Source
> smkinven           00000000004DD153  wrpdemis_                 284  wrpdemis.f
> smkinven           000000000041EB22  genpdout_                 302  genpdout.f
> smkinven           00000000004AD513  MAIN__                    502  smkinven.f
> smkinven           00000000004072DD  Unknown               Unknown  Unknown
> libc-2.28.so       0000152D846CDD85  __libc_start_main     Unknown  Unknown
> smkinven           00000000004071FE  Unknown               Unknown  Unknown
> 31.397u 0.353s 0:32.17 98.6%    0+0k 464+104328io 4pf+0w
> ```
> 
> [Code line where issue occurs](https://github.com/CEMPD/SMOKE/blob/97d0e2c759556b2a1da2489cee80814c6954921b/src/smkinven/wrpdemis.f#L284C1-L284C33)
> 
> This issue seems only applicable to _ptegu_ when processing the daily emission processing script (after the onetime processing script was executed).

Three issues were found in subroutine WRPDEMIS that caused array out of bound errors. Two issues were fixed by rearrange variable values when processing CEM data. The 3rd issue was related to treatment of special data variable when processing CEM data: variable CODEA record special data 9006 which represents FLOWPOS (position of flow rate in CEM data format). CODEA variable carries indexes to inventory pollutant names (EANAM), and when value 9006 was called as index of EANAM, out-of-bound error occurred.

Code lines 271 - 316 are currently not well constructed with many value re-assignments of V and POLNAM variables when processing CEMS.
